README
FireHOL, an iptables stateful packet filtering firewall for humans!
FireQOS, a TC based bandwidth shaper for humans!

                        http://firehol.org/


Upgrade Notes
=============
From version 2.0.0-pre6, FireHOL adds combined IPv4/IPv6 support within
a single configuration.

If you are upgrading FireHOL from a version earlier than 2.0.0-pre6,
please read the upgrade notes, here:
   http://firehol.org/upgrade/


Installation
============
If you are installing direct from GitHub, please see:
  http://firehol.org/source-install/

If you are installing the package from a tar-files release, FireHOL uses
the GNU Autotools so you can get away with:
  ./configure && make && make install

You can set the compilation options for iprange like this:
  CFLAGS=-O3 ./configure && make && make install

Take a look at the file INSTALL, also.


Getting Started
===============
Configuration for FireHOL goes in /etc/firehol/firehol.conf
Configuration for FireQOS goes in /etc/firehol/fireqos.conf

In the examples directory, you can find examples for both programs.

To start the programs:
  firehol start
  fireqos start

For more details on the command-line options, see the man-pages:
  man firehol
  man fireqos

Read the tutorials on the website for more information and to learn
how to configure the programs:
  http://firehol.org/tutorial/

For detailed information on the configuration files, read the manual
online, or start with these the man-pages:
  man firehol.conf
  man fireqos.conf

You may want to ensure that FireHOL and FireQOS run at boot-time. If you
installed from an distribution package this will be configured in the
usual way.

For a tar-file installation, the binaries can often be linked directly
into /etc/init.d, since their options are SysVInit compatible. A systemd
service file for FireHOL can be found here:
  https://github.com/ktsaou/firehol/issues/17


Support and documentation
=========================
The main website is here:
   http://firehol.org/

To ask questions please sign up to the list:
   http://lists.firehol.org/mailman/listinfo/firehol-support

Man pages, PDF and HTML documentation are provided as part of the package
and can be found in the tarball or in your distribution's standard locations
(e.g. /usr/share/doc). The latest manual is always available here:
   http://firehol.org/manual/

The list of all services supported by FireHOL "out of the box" and
information on adding new services is here:
   http://firehol.org/services/


License
=======
Copyright (C) 2012,2013 Phil Whineray <phil@sanewall.org>
Copyright (C) 2003-2013 Costa Tsaousis <costa@tsaousis.gr>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA


NEWS
2016/01/10:

Release of version 2.0.4

- Add ipv6mld to simplify enabling Multicast Listener Discovery protocol,
  required on networks which do multicast snooping.
- Update the example to make it more likely to work copy-pasted, include MLD

2015/04/26:

Release of version 2.0.3

- Fix to avoid errors when using physin/physout. Note that these
  parameters are only useful when the traffic travels over a single
  bridge.
- Fix to handle tc output on some systems, courtesy of Phineas Gage

2015/03/06:

Release of version 2.0.2

- Fix to ensure the final firewall close code emits as both ipv4 and ipv6
  where appropriate even if only ipv4 or ipv6 was used for the final
  interface/router
- Added --disable-doc to configure script to stop the installation
  of PDF and HTML versions of documentation

2015/02/15:

Release of version 2.0.1

- Fix srcmac/dstmac with IPv6
- Stop configs embedding iptables -I statements becoming corrupted.

  Fixes #41 by deleting activation rules by spec not number.

2014/10/24:

Release of version 2.0.0

No changes compared to rc.3

2014/10/19:

Release of version 2.0.0-rc.3

- Fix chain lengths and ensure both IPv4/IPv6 ones created for
  "with limit", "with knock" and "with recent" (issues #38 and #40)
- Silently disable IPv6 where the kernel has no IPv6 support (#39)

2014/10/06:

Release of version 2.0.0-rc.2

FireHOL fixes/enhancements
- Create functional firehol helpme output (issue #35)
- Remove long-redundant firehol_wget and wget_cmd helpers
- Use mktemp for temporary directories during RPC enumeration
- Don't delete and recreate the main temporary directory
- Treat mktemp like other required commands
- Silence module detection warning when not loading modules

FireQOS fixes/enhancements
- Added srcmac dstmac matches to FireQOS

Packaging
- Clean up some intermediate files before packing

2014/08/02:

Release of version 2.0.0-rc.1

FireHOL fixes/enhancements
- #15 Do not fix source ports for DHCPv6 (not required by RFC)

FireQOS fixes/enhancements
- fixed mixed ipv4 and ipv6 matches that were generating match
  priorities above 0xffff
- #29 added warning when rate is lower than minrate

Documentation
- split manual in two
- explain that ICMPv6 ND/RD packets are untracked by default, so not in all
- #31 fixed magic line of services definition

2014/06/08:

Release of version 2.0.0-pre9

FireHOL fixes/enhancements
- recognise dst4/dst6 on interface/router as well as src4/src6

2014/05/10:

Release of version 2.0.0-pre8

FireQOS fixes/enhancements
- resolved an issue with pppd ip-up scripts
- automatic numbering continues giving class priorities after
  a manual priority is given

Documentation
- replaced docbook with pandoc (issue #24)

2014/04/13:

Release of version 2.0.0-pre7

FireQOS fixes/enhancements
- full bidirectional interface support, including firehol like services
- running on OpenWRT (insmod instead of modprobe, low-res timers)
- option to limit each match to a specific rate

FireHOL fixes/enhancements
- firehol save now writes to the specified files again
- firehol save, restore and fastactivation now work IPv4-only mode
- improved fast activation error handling (issue #22)
- improved mark/connmark handling (issue #23)
- tproxy support added

Documentation
    - Link and spelling fixes
    - new config examples

2014/02/15:

Release of version 2.0.0-pre6

Implement dual IPv4/IPv6 as standard
- Update your config from version 5 to 6 to enable
- See http://firehol.org/upgrade/#config-version-6

Include FireQOS manual pages

2013/10/29:

Release of version 2.0.0-pre5

Fix to the FIREHOL_DEBUGGING code.

Release of version 2.0.0-pre4

Cleanups and minor improvements
- Remove the blank line from wizard/helpme output
- Remove the #! from wizard output - it cannot be executed directly
- Cleaned up boilerplace information in script
- "firehol condrestart" now follows official conventions by restarting
   only if already running

Fix kernel version detection so that it is more flexible and less error prone.

Allow switching on of debug output by setting environment variable
FIREHOL_DEBUGGING to a non-empty value.

Fix "mac" helper command so that it works with iptables 1.4.12+
- Previous behaviour was deprecated in 1.4.3 (Jul 2009)
- Also prevent mac addresses being seen as IPv6 addresses

Allow some config variables to be set as environment variables
- FIREHOL_ESTABLISHED_ACTIVATION_ACCEPT,
  FIREHOL_INPUT_ACTIVATION_POLICY, FIREHOL_FORWARD_ACTIVATION_POLICY,
  FIREHOL_OUTPUT_ACTIVATION_POLICY, FIREHOL_LOAD_KERNEL_MODULES,
  FIREHOL_NAT, FIREHOL_ROUTING, FIREHOL_AUTOSAVE

Do not try to add DROP rules to NAT chains
- See http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=536675

Use "ss" in place of "netstat"
- As suggested here https://bugzilla.redhat.com/show_bug.cgi?id=784520

2013/10/28:

Release of version 2.0.0-pre3

Fix for issue https://github.com/ktsaou/firehol/issues/6
Fix for issue https://github.com/ktsaou/firehol/issues/9

2013/10/27:

Release of version 2.0.0-pre2

Inclusion of FireQOS
Standardisation of version information suitable for Git
Allow tcpmss usage in interfaces

2013/10/15:

Release of version 2.0.0-pre1

Project organisation moved to autotools: it is still possible to
copy the single sbin/firehol.in to an appropriate place on an
init-driven system.


get-iana.sh
#!/bin/bash

# $Id: get-iana.sh,v 1.15 2013/01/06 23:49:08 ktsaou Exp $
#
# $Log: get-iana.sh,v $
# Revision 1.15  2013/01/06 23:49:08  ktsaou
# Removed depedency to get-iana.sh
# It is not usefull any more.
#
# Revision 1.14  2010/06/07 15:44:09  ktsaou
# Made get-iana.sh support the latest IANA format.
#
# Revision 1.13  2010/04/08 22:03:08  ktsaou
# Removed --proxy=off for wget.
#
# Revision 1.12  2008/03/17 22:08:43  ktsaou
# Updated for latest IANA reservations format.
#
# Revision 1.11  2007/06/13 14:40:04  ktsaou
# *** empty log message ***
#
# Revision 1.10  2007/05/05 23:38:31  ktsaou
# Added support for external definitions of:
#
# RESERVED_IPS
# PRIVATE_IPS
# MULTICAST_IPS
# UNROUTABLE_IPS
#
# in files under the same name in /etc/firehol/.
# Only RESERVED_IPS is mandatory (firehol will complain if it is not there,
# but it will still work without it), and is also the only file that firehol
# checks how old is it. If it is 90+ days old, firehol will complain again.
#
# Changed the supplied get-iana.sh script to generate the RESERVED_IPS file.
# FireHOL also instructs the user to use this script if the file is missing
# or is too old.
#
# Revision 1.9  2007/04/29 19:34:11  ktsaou
# *** empty log message ***
#
# Revision 1.8  2005/06/02 15:48:52  ktsaou
# Allowed 127.0.0.1 to be in RESERVED_IPS
#
# Revision 1.7  2005/05/08 23:27:23  ktsaou
# Updated RESERVED_IPS to current IANA reservations.
#
# Revision 1.6  2004/01/10 18:44:39  ktsaou
# Further optimized and reduced PRIVATE_IPS using:
# http://www.vergenet.net/linux/aggregate/
#
# The supplied get-iana.sh uses 'aggregate-flim' if it finds it in the path.
# (aggregate-flim is the name of this program when installed on Gentoo)
#
# Revision 1.5  2003/08/23 23:26:50  ktsaou
# Bug #793889:
# Change #!/bin/sh to #!/bin/bash to allow FireHOL run on systems that
# bash is not linked to /bin/sh.
#
# Revision 1.4  2002/10/27 12:44:42  ktsaou
# CVS test
#

#
# Program that downloads the IPv4 address space allocation by IANA
# and creates a list with all reserved address spaces.
#

# IPV4_ADDRESS_SPACE_URL="http://www.iana.org/assignments/ipv4-address-space"
IPV4_ADDRESS_SPACE_URL="http://www.iana.org/assignments/ipv4-address-space/ipv4-address-space.txt"

# The program will match all rows in the file which start with a number, have a slash,
# followed by another number, for which the following pattern will also match on the
# same rows
IANA_RESERVED="(RESERVED|UNALLOCATED)"

# which rows that are matched by the above, to ignore
# (i.e. not include them in RESERVED_IPS)?
#IANA_IGNORE="(Multicast|Private use|Loopback|Local Identification)"
IANA_IGNORE="Multicast"

tempfile="/tmp/iana.$$.$RANDOM"

AGGREGATE="`which aggregate-flim 2>/dev/null`"
if [ -z "${AGGREGATE}" ]
then
	AGGREGATE="`which aggregate 2>/dev/null`"
fi

if [ -z "${AGGREGATE}" ]
then
	echo >&2
	echo >&2
	echo >&2 "WARNING"
	echo >&2 "Please install 'aggregate-flim' to shrink the list of IPs."
	echo >&2
	echo >&2
fi

echo >&2
echo >&2 "Fetching IANA IPv4 Address Space, from:"
echo >&2 "${IPV4_ADDRESS_SPACE_URL}"
echo >&2

wget -O - "${IPV4_ADDRESS_SPACE_URL}"	|\
	egrep "^ *[0-9]+/[0-9]+.*${IANA_RESERVED}"	|\
	egrep -vi "${IANA_IGNORE}"			|\
	sed "s/^ \+//g"					|\
	cut -d ' ' -f 1					|\
(
	while IFS="/" read range net
	do
		# echo >&2 "$range/$net"
		
		if [ ! $net -eq 8 ]
		then
			echo >&2 "Cannot handle network masks of $net bits ($range/$net)"
			continue
		fi
		 
		first=`echo $range | cut -d '-' -f 1`
		first=`expr $first + 0`
		last=`echo $range | cut -d '-' -f 2`
		last=`expr $last + 0`
		
		x=$first
		while [ ! $x -gt $last ]
		do
			# test $x -ne 127 && echo "$x.0.0.0/$net"
			echo "$x.0.0.0/$net"
			x=$[x + 1]
		done
	done
) | \
(
	if [ ! -z "${AGGREGATE}" -a -x "${AGGREGATE}" ]
	then
		"${AGGREGATE}"
	else
		cat
	fi
) >"${tempfile}"

echo >&2 
echo >&2 
echo >&2 "FOUND THE FOLLOWING RESERVED IP RANGES:"
printf "RESERVED_IPS=\""
i=0
for x in `cat ${tempfile}`
do
	i=$[i + 1]
	printf "${x} "
done
printf "\"\n"

if [ $i -eq 0 ]
then
	echo >&2 
	echo >&2 
	echo >&2 "Failed to find reserved IPs."
	echo >&2 "Possibly the file format has been changed, or I cannot fetch the URL."
	echo >&2 
	
	rm -f ${tempfile}
	exit 1
fi
echo >&2
echo >&2
echo >&2 "Differences between the fetched list and the list installed in"
echo >&2 "/etc/firehol/RESERVED_IPS:"

echo >&2 "# diff /etc/firehol/RESERVED_IPS ${tempfile}"
diff /etc/firehol/RESERVED_IPS ${tempfile}

if [ $? -eq 0 ]
then
	touch /etc/firehol/RESERVED_IPS
	echo >&2
	echo >&2 "No differences found."
	echo >&2
	
	rm -f ${tempfile}
	exit 0
fi

echo >&2 
echo >&2 
echo >&2 "Would you like to save this list to /etc/firehol/RESERVED_IPS"
echo >&2 "so that FireHOL will automatically use it from now on?"
echo >&2
while [ 1 = 1 ]
do
	printf >&2 "yes or no > "
	read x
	
	case "${x}" in
		yes)	cp -f /etc/firehol/RESERVED_IPS /etc/firehol/RESERVED_IPS.old 2>/dev/null
			cat "${tempfile}" >/etc/firehol/RESERVED_IPS || exit 1
			echo >&2 "New RESERVED_IPS written to '/etc/firehol/RESERVED_IPS'."
			break
			;;
			
		no)
			echo >&2 "Saved nothing."
			break
			;;
			
		*)	echo >&2 "Cannot understand '${x}'."
			;;
	esac
done

rm -f ${tempfile}

